README.md
# day-schedule-selector [![Build Status](https://travis-ci.org/artsy/day-schedule-selector.svg?branch=master)](https://travis-ci.org/artsy/day-schedule-selector)

A jQuery plugin to render a weekly schedule and allow selecting time slots in each day.

![demo](https://cloud.githubusercontent.com/assets/796573/7264504/23c2109a-e85a-11e4-9c26-19358686cbb0.gif)

## Getting Started
```javascript
$('#weekly-schedule').dayScheduleSelector({
  /* options */
});
```
## Options

```
$("#weekly-schedule").dayScheduleSelector({
  days: [1, 2, 3, 5, 6],
  startTime: '09:50',
  endTime: '21:06',
  interval: 15
});
```

#### days
default: `[0, 1, 2, 3, 4, 5, 6]`

An integer array of days included in the calendar with Sunday being `0` and Saturday being `6`.

#### startTime
default: `'08:00'`

Start time of each day on the calendar, in `HH:mm` format.

#### endTime
default: `'20:00'`

End time of each day on the calendar, in `HH:mm` format.

#### interval
default: `30`

An integer value representing length of each time slot, in minutes.

## Events
The following custom events are triggered on the element.

#### selected.artsy.dayScheduleSelector
Triggered when a selection is made. Passes the event and an array of selected time slots to the event handler.
```javascript
$("#weekly-schedule").on('selected.artsy.dayScheduleSelector', function (e, selected) {
  /* selected is an array of time slots selected this time. */
}
```

#### deselected.artsy.dayScheduleSelector
Triggered when deselecting a slot. Passes the event and the slot that is being deselected.
```javascript
$("#weekly-schedule").on('deselected.artsy.dayScheduleSelector', function (e, deselected) {
  /* To access to the attributes 'data-day' and 'data-time'*/
  var dataDay = deselected[0].getAttribute("data-day");
  var dataTime = deselected[0].getAttribute("data-time");
}
```

## Installation

### Install as a Ruby gem

Add this line to your application's Gemfile:

```ruby
gem 'day_schedule_selector', '0.1.0', { git: 'https://github.com/starsirius/day-schedule-selector.git', branch: 'master' }
```

And then execute:

    $ bundle

## Usage

TODO: Write usage instructions here

## Development

### Ruby gem

After checking out the repo, run `bin/setup` to install dependencies. Then, run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release` to create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

1. Fork it ( https://github.com/starsirius/day-schedule-selector/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create a new Pull Request


CHANGELOG.md
(Next)
==================

#### Features

* [#33](https://github.com/artsy/day-schedule-selector/pull/33): Add Danger; add mocha, chai, and sinon to dev dependencies - [@starsirius](https://github.com/starsirius).

#### Fixes

* [#35](https://github.com/artsy/day-schedule-selector/pull/35): Fix empty last time slot when there are many calendars  - [@adelriosantiago](https://github.com/adelriosantiago).

test/index.js
"use strict";

describe ("DayScheduleSelector", function () {
  beforeEach (function () {
    document.body.innerHTML = '<div id="schedule-selector"></div>';
  });

  describe ("initialization", function () {
    describe ("with default options", function () {
      beforeEach (function () {
        $("#schedule-selector").dayScheduleSelector();
      });

      it ("initializes the day labels properly", function() {
        var stringDays = $("#schedule-selector").data('artsy.dayScheduleSelector').options.stringDays
          , days = $("#schedule-selector").data('artsy.dayScheduleSelector').options.days;

        $(".schedule-header th").length.should.eq(days.length + 1);
        $.each(days, function(index, day) {
          $(".schedule-header th").eq(index + 1).text().should.eq(stringDays[index] || "");
        });
      });

      it ("initializes the time labels properly");

      it ("initializes the time slots properly", function () {
        $("#schedule-selector .time-slot").length.should.eq(24 * 7);
      });
    });

    describe ("with custom options", function () {
      beforeEach (function () {
        $("#schedule-selector").dayScheduleSelector({
          days: [1, 3, 5, 6],
          stringDays: ['星期一', '星期三', '星期五'],
          startTime: '12:00',
          endTime: '17:00',
          interval: 20
        });
      })

      it ("initializes the day labels properly", function() {
        var stringDays = $("#schedule-selector").data('artsy.dayScheduleSelector').options.stringDays
          , days = $("#schedule-selector").data('artsy.dayScheduleSelector').options.days;

        $(".schedule-header th").length.should.eq(days.length + 1);
        $.each(days, function(index, day) {
          $(".schedule-header th").eq(index + 1).text().should.eq(stringDays[index] || "");
        });
      });

      it ("initializes the time labels properly");

      it ("initializes the time slots properly");
    });
  });

  describe ("selection", function () {
    beforeEach (function () {
      $("#schedule-selector").dayScheduleSelector();
    });

    describe ("clicking on the time slots", function () {
      it ("selects the time slot after the first click", function () {
        var $slot = $("#schedule-selector .time-slot").first();

        $slot.click()
        $slot.is("[data-selecting='selecting']").should.be.ok;
        $("#schedule-selector").data('artsy.dayScheduleSelector').isSelecting().should.be.ok;
      });

      it ("disables other days after the first click");
    });

    $.each([".time-label", "th"], function (i, s) {
      describe ("clicking on " + s, function () {
        it ("does not select anything", function() {
          $("#schedule-selector").find(s).length.should.be.above(0);
          $("#schedule-selector").find(s).click();
          $("#schedule-selector").data('artsy.dayScheduleSelector').isSelecting().should.not.be.ok
          $("#schedule-selector").find(s).each(function (_, el) {
            $(el).is("[data-selecting], [data-selected]").should.not.be.ok
          });
        });
      });
    });
  });

  describe ("events", function () {
    beforeEach (function () {
      $("#schedule-selector").dayScheduleSelector();
    });

    describe ("selected", function () {
      it ("receives the selected event properly with correct data", function (done) {
        $("#schedule-selector").on('selected.artsy.dayScheduleSelector', function  (e, $selected) {
          $selected.length.should.eq(5);
          done();
        });
        $("#schedule-selector .time-slot[data-time='08:30'][data-day='3']").click();
        $("#schedule-selector .time-slot[data-time='10:30'][data-day='3']").click();
      });
    })
  });

  describe ("#serialize", function() {
    beforeEach (function() {
      var selections = [
        [0, '08:00'], [0, '19:30'], // first one and last one
        [1, '08:00'], [1, '08:30'], [1, '19:00'], [1, '19:30'], // first two and last two
        [2, '19:30'], // only last one
        // all day for 3 (Wednesday),
        [4, '08:00'], [4, '09:00'], [4, '10:00'], [4, '11:00'], [4, '12:00'], // every other slots in the morning
        [5, '13:00'], [5, '14:00'], [5, '15:00'], [5, '16:00'], [5, '17:00'], // every other slots in the afternoon
        [6, '08:00'], [6, '08:30'], [6, '09:00'], [6, '11:30'], [6, '12:00'], [6, '18:30'], [6, '19:00'], [6, '19:30']
      ]
      $("#schedule-selector").dayScheduleSelector();
      for (var i = 0; i < selections.length; i++) {
        $("#schedule-selector .time-slot[data-day='" + selections[i][0] +"'][data-time='" + selections[i][1] + "']").attr('data-selected', 'selected');
      }
      // Select all day for Wednesday
      $("#schedule-selector .time-slot[data-day='3']").attr('data-selected', 'selected');
    });

    it ("serializes the selected time slots correctly", function () {
      var selected = $("#schedule-selector").data('artsy.dayScheduleSelector').serialize()
      selected.should.eql({
        0: [['08:00', '08:30'], ['19:30', '20:00']],
        1: [['08:00', '09:00'], ['19:00', '20:00']],
        2: [['19:30', '20:00']],
        3: [['08:00', '20:00']],
        4: [['08:00', '08:30'], ['09:00', '09:30'], ['10:00', '10:30'], ['11:00', '11:30'], ['12:00', '12:30']],
        5: [['13:00', '13:30'], ['14:00', '14:30'], ['15:00', '15:30'], ['16:00', '16:30'], ['17:00', '17:30']],
        6: [['08:00', '09:30'], ['11:30', '12:30'], ['18:30', '20:00']]
      });
    });
  });
});


karma.conf.js
// Karma configuration
// Generated on Wed Apr 29 2015 16:48:15 GMT-0400 (EDT)

module.exports = function(config) {
  config.set({

    // base path that will be used to resolve all patterns (eg. files, exclude)
    basePath: '',


    // frameworks to use
    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
    frameworks: ['mocha', 'chai', 'sinon'],

    // list of files / patterns to load in the browser
    files: [
      'examples/vendor/javascripts/jquery-1.11.2.js',
      'src/index.js',
      'test/*.js'
    ],


    // list of files to exclude
    exclude: [
    ],


    // preprocess matching files before serving them to the browser
    // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
    preprocessors: {
    },


    // test results reporter to use
    // possible values: 'dots', 'progress'
    // available reporters: https://npmjs.org/browse/keyword/karma-reporter
    reporters: ['mocha'],


    // web server port
    port: 9876,


    // enable / disable colors in the output (reporters and logs)
    colors: true,


    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_INFO,


    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: true,


    // start these browsers
    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
    browsers: ['PhantomJS'],


    // Continuous Integration mode
    // if true, Karma captures browsers, runs the tests and exits
    singleRun: true
  });
};


src/index.js
(function ($) {
  'use strict';

  var DayScheduleSelector = function (el, options) {
    this.$el = $(el);
    this.options = $.extend({}, DayScheduleSelector.DEFAULTS, options);
    this.render();
    this.attachEvents();
    this.$selectingStart = null;
  }

  DayScheduleSelector.DEFAULTS = {
    days        : [0, 1, 2, 3, 4, 5, 6],  // Sun - Sat
    startTime   : '08:00',                // HH:mm format
    endTime     : '20:00',                // HH:mm format
    interval    : 30,                     // minutes
    stringDays  : ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'],
    template    : '<div class="day-schedule-selector">'         +
                    '<table class="schedule-table">'            +
                      '<thead class="schedule-header"></thead>' +
                      '<tbody class="schedule-rows"></tbody>'   +
                    '</table>'                                  +
                  '<div>'
  };

  /**
   * Render the calendar UI
   * @public
   */
  DayScheduleSelector.prototype.render = function () {
    this.$el.html(this.options.template);
    this.renderHeader();
    this.renderRows();
  };

  /**
   * Render the calendar header
   * @public
   */
  DayScheduleSelector.prototype.renderHeader = function () {
    var stringDays = this.options.stringDays
      , days = this.options.days
      , html = '';

    $.each(days, function (i, _) { html += '<th>' + (stringDays[i] || '') + '</th>'; });
    this.$el.find('.schedule-header').html('<tr><th></th>' + html + '</tr>');
  };

  /**
   * Render the calendar rows, including the time slots and labels
   * @public
   */
  DayScheduleSelector.prototype.renderRows = function () {
    var start = this.options.startTime
      , end = this.options.endTime
      , interval = this.options.interval
      , days = this.options.days
      , $el = this.$el.find('.schedule-rows');

    $.each(generateDates(start, end, interval), function (i, d) {
      var daysInARow = $.map(new Array(days.length), function (_, i) {
        return '<td class="time-slot" data-time="' + hhmm(d) + '" data-day="' + days[i] + '"></td>'
      }).join();

      $el.append('<tr><td class="time-label">' + hmmAmPm(d) + '</td>' + daysInARow + '</tr>');
    });
  };

  /**
   * Is the day schedule selector in selecting mode?
   * @public
   */
  DayScheduleSelector.prototype.isSelecting = function () {
    return !!this.$selectingStart;
  }

  DayScheduleSelector.prototype.select = function ($slot) { $slot.attr('data-selected', 'selected'); }
  DayScheduleSelector.prototype.deselect = function ($slot) { $slot.removeAttr('data-selected'); }

  function isSlotSelected($slot) { return $slot.is('[data-selected]'); }
  function isSlotSelecting($slot) { return $slot.is('[data-selecting]'); }

  /**
   * Get the selected time slots given a starting and a ending slot
   * @private
   * @returns {Array} An array of selected time slots
   */
  function getSelection(plugin, $a, $b) {
    var $slots, small, large, temp;
    if (!$a.hasClass('time-slot') || !$b.hasClass('time-slot') ||
        ($a.data('day') != $b.data('day'))) { return []; }
    $slots = plugin.$el.find('.time-slot[data-day="' + $a.data('day') + '"]');
    small = $slots.index($a); large = $slots.index($b);
    if (small > large) { temp = small; small = large; large = temp; }
    return $slots.slice(small, large + 1);
  }

  DayScheduleSelector.prototype.attachEvents = function () {
    var plugin = this
      , options = this.options
      , $slots
      , purpose;

    this.$el.on('click', '.time-slot', function () {
      var day = $(this).data('day');
      if (!plugin.isSelecting()) {  // if we are not in selecting mode
        purpose = isSlotSelected($(this)) ? 'deselecting' : 'selecting';

          // start selecting or deselecting
          plugin.$selectingStart = $(this);
          $(this).attr('data-selecting', purpose)
          plugin.$el.find('.time-slot').attr('data-disabled', 'disabled');
          plugin.$el.find('.time-slot[data-day="' + day + '"]').removeAttr('data-disabled');
      } else {  // if we are in selecting mode
        if (day == plugin.$selectingStart.data('day')) {  // if clicking on the same day column
          // then end of selection
          if(purpose === 'selecting'){
            plugin.$el.find('.time-slot[data-day="' + day + '"]').filter('[data-selecting]')
            .attr('data-selected', 'selected').removeAttr('data-selecting');
          } else if(purpose === 'deselecting'){
            plugin.$el.find('.time-slot[data-day="' + day + '"]').filter('[data-selecting]')
            .removeAttr('data-selected').removeAttr('data-selecting');
          }


          plugin.$el.find('.time-slot').removeAttr('data-disabled');
          plugin.$el.trigger('selected.artsy.dayScheduleSelector', [getSelection(plugin, plugin.$selectingStart, $(this))]);
          plugin.$selectingStart = null;
        }
      }
    });

    this.$el.on('mouseover', '.time-slot', function () {
      var $slots, day, start, end, temp;
      if (plugin.isSelecting()) {  // if we are in selecting mode
        day = plugin.$selectingStart.data('day');
        $slots = plugin.$el.find('.time-slot[data-day="' + day + '"]');
        $slots.filter('[data-selecting]').removeAttr('data-selecting');
        start = $slots.index(plugin.$selectingStart);
        end = $slots.index(this);
        if (end < 0) return;  // not hovering on the same column
        if (start > end) { temp = start; start = end; end = temp; }
        $slots.slice(start, end + 1).attr('data-selecting', purpose)

      }
    });
  };

  /**
   * Serialize the selections
   * @public
   * @returns {Object} An object containing the selections of each day, e.g.
   *    {
   *      0: [],
   *      1: [["15:00", "16:30"]],
   *      2: [],
   *      3: [],
   *      5: [["09:00", "12:30"], ["15:00", "16:30"]],
   *      6: []
   *    }
   */
  DayScheduleSelector.prototype.serialize = function () {
    var plugin = this
      , selections = {};

    $.each(this.options.days, function (_, v) {
      var start, end;
      start = end = false; selections[v] = [];
      plugin.$el.find(".time-slot[data-day='" + v + "']").each(function () {
        // Start of selection
        if (isSlotSelected($(this)) && !start) {
          start = $(this).data('time');
        }

        // End of selection (I am not selected, so select until my previous one.)
        if (!isSlotSelected($(this)) && !!start) {
          end = $(this).data('time');
        }

        // End of selection (I am the last one :) .)
        if (isSlotSelected($(this)) && !!start && $(this).is($(this).closest('tbody').find(".time-slot[data-day='" + v + "']:last"))) {
          end = secondsSinceMidnightToHhmm(
            hhmmToSecondsSinceMidnight($(this).data('time')) + plugin.options.interval * 60);
        }

        if (!!end) { selections[v].push([start, end]); start = end = false; }
      });
    })
    return selections;
  };

  /**
   * Deserialize the schedule and render on the UI
   * @public
   * @param {Object} schedule An object containing the schedule of each day, e.g.
   *    {
   *      0: [],
   *      1: [["15:00", "16:30"]],
   *      2: [],
   *      3: [],
   *      5: [["09:00", "12:30"], ["15:00", "16:30"]],
   *      6: []
   *    }
   */
  DayScheduleSelector.prototype.deserialize = function (schedule) {
    var plugin = this, i;
    $.each(schedule, function(d, ds) {
      var $slots = plugin.$el.find('.time-slot[data-day="' + d + '"]');
      $.each(ds, function(_, s) {
        for (i = 0; i < $slots.length; i++) {
          if ($slots.eq(i).data('time') >= s[1]) { break; }
          if ($slots.eq(i).data('time') >= s[0]) { plugin.select($slots.eq(i)); }
        }
      })
    });
  };

  // DayScheduleSelector Plugin Definition
  // =====================================

  function Plugin(option) {
    return this.each(function (){
      var $this   = $(this)
        , data    = $this.data('artsy.dayScheduleSelector')
        , options = typeof option == 'object' && option;

      if (!data) {
        $this.data('artsy.dayScheduleSelector', (data = new DayScheduleSelector(this, options)));
      }
    })
  }

  $.fn.dayScheduleSelector = Plugin;

  /**
   * Generate Date objects for each time slot in a day
   * @private
   * @param {String} start Start time in HH:mm format, e.g. "08:00"
   * @param {String} end End time in HH:mm format, e.g. "21:00"
   * @param {Number} interval Interval of each time slot in minutes, e.g. 30 (minutes)
   * @returns {Array} An array of Date objects representing the start time of the time slots
   */
  function generateDates(start, end, interval) {
    var numOfRows = Math.ceil(timeDiff(start, end) / interval);
    return $.map(new Array(numOfRows), function (_, i) {
      // need a dummy date to utilize the Date object
      return new Date(new Date(2000, 0, 1, start.split(':')[0], start.split(':')[1]).getTime() + i * interval * 60000);
    });
  }

  /**
   * Return time difference in minutes
   * @private
   */
  function timeDiff(start, end) {   // time in HH:mm format
    // need a dummy date to utilize the Date object
    return (new Date(2000, 0, 1, end.split(':')[0], end.split(':')[1]).getTime() -
            new Date(2000, 0, 1, start.split(':')[0], start.split(':')[1]).getTime()) / 60000;
  }

  /**
   * Convert a Date object to time in H:mm format with am/pm
   * @private
   * @returns {String} Time in H:mm format with am/pm, e.g. '9:30am'
   */
  function hmmAmPm(date) {
    var hours = date.getHours()
      , minutes = date.getMinutes()
      , ampm = hours >= 12 ? 'pm' : 'am';
    return hours + ':' + ('0' + minutes).slice(-2) + ampm;
  }

  /**
   * Convert a Date object to time in HH:mm format
   * @private
   * @returns {String} Time in HH:mm format, e.g. '09:30'
   */
  function hhmm(date) {
    var hours = date.getHours()
      , minutes = date.getMinutes();
    return ('0' + hours).slice(-2) + ':' + ('0' + minutes).slice(-2);
  }

  function hhmmToSecondsSinceMidnight(hhmm) {
    var h = hhmm.split(':')[0]
      , m = hhmm.split(':')[1];
    return parseInt(h, 10) * 60 * 60 + parseInt(m, 10) * 60;
  }

  /**
   * Convert seconds since midnight to HH:mm string, and simply
   * ignore the seconds.
   */
  function secondsSinceMidnightToHhmm(seconds) {
    var minutes = Math.floor(seconds / 60);
    return ('0' + Math.floor(minutes / 60)).slice(-2) + ':' +
           ('0' + (minutes % 60)).slice(-2);
  }

  // Expose some utility functions
  window.DayScheduleSelector = {
    ssmToHhmm: secondsSinceMidnightToHhmm,
    hhmmToSsm: hhmmToSecondsSinceMidnight
  };

})(jQuery);
